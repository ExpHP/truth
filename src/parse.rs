use crate::lalrparser;
use crate::{ast, meta};
use crate::lexer::{Lexer, Token};

pub type ErrorPayload = &'static str; // FIXME: this is the default for LALRPOP
pub type Error<'input> = lalrpop_util::ParseError<usize, Token<'input>, ErrorPayload>;
pub type Result<'input, T> = std::result::Result<T, Error<'input>>;

pub trait Parse<'input>: Sized {
    fn parse<B: AsRef<[u8]> + ?Sized>(s: &'input B) -> Result<'input, Self> {
        Self::parse_stream(Lexer::new(s.as_ref()))
    }

    fn parse_stream(lexer: Lexer<'input>) -> Result<'input, Self>;
}

/// Return type of the LALRPOP `Anything` parser.
///
/// This is only an implementation detail of the [`Parse`] trait.
/// Use that instead.
#[doc(hidden)]
pub enum AnythingValue {
    Script(ast::Script),
    Item(ast::Item),
    Stmt(ast::Stmt),
    Expr(ast::Expr),
    Var(ast::Var),
    Ident(ast::Ident),
    Meta(meta::Meta),
    LitString(ast::LitString),
}

/// Implementation detail of the [`Parse`] trait.  Use that instead.
#[doc(hidden)]
#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum AnythingTag {
    Script,
    Item,
    Stmt,
    Expr,
    Var,
    Ident,
    LitString,
    Meta,
}

// Inserts the virtual token and calls the Anything parser
fn call_anything_parser(tag: AnythingTag, lexer: Lexer<'_>) -> Result<'_, AnythingValue> {
    let offset = lexer.offset();
    let lexer = std::iter::once(Ok((offset, Token::VirtualDispatch(tag), offset))).chain(lexer);
    lalrparser::AnythingParser::new().parse(lexer)
}

impl<'input> Parse<'input> for ast::Script {
    fn parse_stream(lexer: Lexer<'input>) -> Result<'input, Self> {
        match call_anything_parser(AnythingTag::Script, lexer)? {
            AnythingValue::Script(x) => Ok(x),
            _ => unreachable!(),
        }
    }
}

impl<'input> Parse<'input> for ast::Item {
    fn parse_stream(lexer: Lexer<'input>) -> Result<'input, Self> {
        match call_anything_parser(AnythingTag::Item, lexer)? {
            AnythingValue::Item(x) => Ok(x),
            _ => unreachable!(),
        }
    }
}

impl<'input> Parse<'input> for ast::Stmt {
    fn parse_stream(lexer: Lexer<'input>) -> Result<'input, Self> {
        match call_anything_parser(AnythingTag::Stmt, lexer)? {
            AnythingValue::Stmt(x) => Ok(x),
            _ => unreachable!(),
        }
    }
}

impl<'input> Parse<'input> for ast::Expr {
    fn parse_stream(lexer: Lexer<'input>) -> Result<'input, Self> {
        match call_anything_parser(AnythingTag::Expr, lexer)? {
            AnythingValue::Expr(x) => Ok(x),
            _ => unreachable!(),
        }
    }
}

impl<'input> Parse<'input> for ast::Var {
    fn parse_stream(lexer: Lexer<'input>) -> Result<'input, Self> {
        match call_anything_parser(AnythingTag::Var, lexer)? {
            AnythingValue::Var(x) => Ok(x),
            _ => unreachable!(),
        }
    }
}

impl<'input> Parse<'input> for ast::Ident {
    fn parse_stream(lexer: Lexer<'input>) -> Result<'input, Self> {
        match call_anything_parser(AnythingTag::Ident, lexer)? {
            AnythingValue::Ident(x) => Ok(x),
            _ => unreachable!(),
        }
    }
}

impl<'input> Parse<'input> for ast::LitString {
    fn parse_stream(lexer: Lexer<'input>) -> Result<'input, Self> {
        match call_anything_parser(AnythingTag::LitString, lexer)? {
            AnythingValue::LitString(x) => Ok(x),
            _ => unreachable!(),
        }
    }
}

impl<'input> Parse<'input> for meta::Meta {
    fn parse_stream(lexer: Lexer<'input>) -> Result<'input, Self> {
        match call_anything_parser(AnythingTag::Meta, lexer)? {
            AnythingValue::Meta(x) => Ok(x),
            _ => unreachable!(),
        }
    }
}
