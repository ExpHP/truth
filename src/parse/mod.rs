use lalrpop_util::lalrpop_mod;

use crate::error::CompileError;
use crate::ast;
use crate::meta;
use crate::pos::{Sp};

lalrpop_mod!(pub lalrparser, "/parse/lalrparser.rs");
mod lalrparser_util;

use lexer::{Lexer, Token};
pub mod lexer;

#[cfg(test)]
mod tests;

pub trait Parse: Sized {
    /// Parse a string into an AST node.
    ///
    /// This is for quick-and-dirty use only; the spans in the output will have incomplete
    /// information and [`crate::Files`] will be unable to locate the corresponding
    /// strings of text. For proper diagnostics you should prefer the helper method
    /// [`crate::pos::NonUtf8Files::parse`] instead.
    fn parse<B: AsRef<str> + ?Sized>(s: &B) -> Result<'_, Self> {
        let mut state = State::new();
        Self::parse_stream(&mut state, Lexer::new(None, s.as_ref()))
            .map(|x| x.value)
    }

    fn parse_stream<'input>(state: &mut State, lexer: Lexer<'input>) -> Result<'input, Sp<Self>>;
}


pub type Error<'input> = lalrpop_util::ParseError<lexer::Location, Token<'input>, CompileError>;
pub type Result<'input, T> = std::result::Result<T, Error<'input>>;


/// Extra state during parsing.
pub struct State {
    mapfiles: Vec<Sp<ast::LitString>>,
    image_sources: Vec<Sp<ast::LitString>>,

    /// When we are parsing instructions, tracks the last time label so that we can produce an
    /// AST with time fields instead of explicit labels.
    ///
    /// A stack is used *as if* we supported nested function definitions.  In practice, the level at
    /// index 0 gets used exclusively when parsing `Stmt` at toplevel, and a level at index 1 gets
    /// used when parsing `Item`s.
    time_stack: Vec<i32>,
}

impl State {
    pub fn new() -> State { State {
        mapfiles: vec![],
        image_sources: vec![],
        time_stack: vec![0],
    }}
}

/// Return type of the LALRPOP `Anything` parser.
///
/// This is only an implementation detail of the [`Parse`] trait.
/// Use that instead.
#[doc(hidden)]
pub enum AnythingValue {
    Script(ast::Script),
    Item(ast::Item),
    Block(ast::Block),
    Stmt(ast::Stmt),
    Expr(ast::Expr),
    Var(ast::Var),
    Meta(meta::Meta),
    LitString(ast::LitString),
}

/// Implementation detail of the [`Parse`] trait.  Use that instead.
#[doc(hidden)]
#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum AnythingTag {
    Script,
    Item,
    Block,
    Stmt,
    Expr,
    Var,
    LitString,
    Meta,
}

// Inserts the virtual token and calls the Anything parser
fn call_anything_parser<'input>(
    tag: AnythingTag,
    state: &mut State,
    lexer: Lexer<'input>,
) -> Result<'input, Sp<AnythingValue>> {
    let start = lexer.location();
    let lexer = std::iter::once(Ok((start, Token::VirtualDispatch(tag), start))).chain(lexer);
    lalrparser::AnythingParser::new().parse(state, lexer)
}


macro_rules! impl_parse {
    ($AstType:ty, $TagName:ident) => {
        impl Parse for $AstType {
            fn parse_stream<'input>(state: &mut State, lexer: Lexer<'input>) -> Result<'input, Sp<Self>> {
                let sp = call_anything_parser(AnythingTag::$TagName, state, lexer)?;
                Ok(sp!(sp.span => match sp.value {
                    AnythingValue::$TagName(x) => x,
                    _ => unreachable!(),
                }))
            }
        }
    }
}

impl_parse!(ast::Script, Script);
impl_parse!(ast::Item, Item);
impl_parse!(ast::Block, Block);
impl_parse!(ast::Stmt, Stmt);
impl_parse!(ast::Expr, Expr);
impl_parse!(ast::Var, Var);
impl_parse!(ast::LitString, LitString);
impl_parse!(meta::Meta, Meta);
