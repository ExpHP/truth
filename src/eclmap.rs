use std::collections::{BTreeMap};

use crate::pos::{Sp, FileId};
use crate::game::{Game, InstrLanguage};
use crate::diagnostic::{RootEmitter, Emitter};
use crate::ident::Ident;
use crate::io::Fs;
use crate::parse::seqmap::{SeqmapRaw, SeqmapRawSection};
use crate::error::{ErrorReported, GatherErrorIteratorExt};

/// FIXME: Rename to Mapfile, how have I not done this already?!  - Exp
#[derive(Debug)]
pub struct Eclmap {
    pub language: InstrLanguage,
    pub ins_names: BTreeMap<i32, Sp<Ident>>,
    pub ins_signatures: BTreeMap<i32, Sp<String>>,
    pub ins_rets: BTreeMap<i32, Sp<String>>,
    pub gvar_names: BTreeMap<i32, Sp<Ident>>,
    pub gvar_types: BTreeMap<i32, Sp<String>>,
    /// For historic reasons, [`InstrLanguage::Timeline`] has dedicated sections.
    /// When these are seen in a file, they will always define things for timelines
    /// instead of [`Self::language`].
    pub timeline_ins_names: BTreeMap<i32, Sp<Ident>>,
    pub timeline_ins_signatures: BTreeMap<i32, Sp<String>>,
}

impl Eclmap {
    pub fn new(language: InstrLanguage) -> Self {
        Eclmap {
            language,
            ins_names: Default::default(),
            ins_signatures: Default::default(),
            ins_rets: Default::default(),
            gvar_names: Default::default(),
            gvar_types: Default::default(),
            /// Legacy `timeline` sections. These will always apply to [`InstrLanguage::Timeline`] instead of [`Self::language`].
            timeline_ins_names: Default::default(),
            timeline_ins_signatures: Default::default(),
        }
    }

    pub fn load(
        path: impl AsRef<std::path::Path>,
        game: Option<Game>,
        emitter: &RootEmitter,
        // Things gets a bit ugly because a map can refer to more maps
        mut read_file: impl FnMut(&std::path::Path) -> Result<(FileId, String), ErrorReported>,
    ) -> Result<Self, ErrorReported> {
        // canonicalize so paths in gamemaps can be interpreted relative to the gamemap path
        let path = path.as_ref();
        let fs = Fs::new(emitter);
        let path = fs.canonicalize(path).map_err(|e| emitter.emit(e))?;

        let (file_id, file_contents) = read_file(&path)?;
        let mut seqmap = SeqmapRaw::parse(file_id, &file_contents, emitter)?;
        if seqmap.magic == "!gamemap" {
            let game = match game {
                Some(game) => game,
                None => return Err(emitter.emit(error!("can't use gamemap because no game was supplied!")))
            };
            let base_dir = path.parent().expect("filename must have parent");
            let game_specific_map_path = Self::resolve_gamemap(base_dir, seqmap, game, emitter)?;
            let (file_id, file_contents) = read_file(&game_specific_map_path)?;
            seqmap = SeqmapRaw::parse(file_id, &file_contents, emitter)?;
            Self::from_seqmap(seqmap, emitter)
        } else {
            // had to put this in an else to satisfy borrow checker _/o\_
            // hope it doesn't code gen twice?
            Self::from_seqmap(seqmap, emitter)
        }
    }

    /// Check the default map directories for a file whose name is `any.{extension}`
    /// and return it if it exists.
    ///
    /// Intended to be used with `Option::or_else` on an optional map path during decompilation.
    pub fn decomp_map_file_from_env(extension: &'static str) -> Option<std::path::PathBuf> {
        std::env::var_os("TRUTH_MAP_PATH")
            .filter(|s| !s.is_empty())
            .into_iter().flat_map(|paths| {
                std::env::split_paths(&paths)
                    .map(|p| p.join(format!("any.{}", extension.trim_start_matches("."))))
                    .filter(|p| p.exists())
                    .take(1).collect::<Vec<_>>()  // stop borrowing
            }).next()
    }

    fn resolve_gamemap(
        base_dir: &std::path::Path,
        seqmap: SeqmapRaw,
        game: Game,
        emitter: &impl Emitter,
    ) -> Result<std::path::PathBuf, ErrorReported> {
        let SeqmapRaw { magic, sections } = seqmap;
        let mut maps = gather_seqmaps(sections, emitter)?;

        let game_files = match maps.remove(&"game_files".to_string()) {
            Some(game_files) => game_files,
            None => return Err(emitter.emit(error!(
                message("no !game_files section in gamemap"),
                primary(magic, "gamemap without !game_files section"),
            ))),
        };
        for (key, _) in maps {
            emitter.emit(warning!(
                message("unrecognized section in gamemap: {:?}", key),
                primary(key, "unrecognized section"),
            )).ignore();
        }
        let rel_path = match game_files.get(&(game.as_number() as i32)) {
            Some(file) => file,
            None => return Err(emitter.emit(error!(
                message("no entry for {}", game.as_str()),
                primary(magic, "gamemap has no entry for {}", game.as_str()),
            ))),
        };

        Ok(base_dir.join(rel_path))
    }

    pub fn from_seqmap(seqmap: SeqmapRaw, emitter: &impl Emitter) -> Result<Eclmap, ErrorReported> {
        let SeqmapRaw { magic, sections } = seqmap;

        let mut maps = gather_seqmaps(sections, emitter)?;

        // NOTE: Experimental.  We have two options for deciding the language:
        //
        //   - Take the language as an argument to this function (since it should be known, anyways).
        //     (this could be better for Ending MSG)
        //   - Decide it from the magic.
        //     (this could facilitate the separation of "timelinemap"s out from eclmaps)
        //
        // Neither is a clear winner at this point in time, but deciding it from the magic is a smaller
        // diff so we do that for now.    - Exp
        let language = match &magic[..] {
            "!eclmap" => InstrLanguage::Ecl,
            "!anmmap" => InstrLanguage::Anm,
            "!stdmap" => InstrLanguage::Std,
            "!msgmap" => InstrLanguage::Msg,
            _ => return Err(emitter.emit(error!(
                message("bad magic: {:?}", magic),
                primary(magic, "bad magic"),
            ))),
        };

        let mut pop_map = |section: &str| maps.remove(&section.to_string()).unwrap_or_else(BTreeMap::new);
        let parse_idents = |section: &str, m: BTreeMap<i32, Sp<String>>| -> Result<BTreeMap<i32, Sp<Ident>>, ErrorReported> {
            emitter.chain_with(|f| write!(f, "section !{}", section), |emitter| {
                m.into_iter().map(|(key, value)| {
                    let ident = value.parse::<Ident>().map_err(|e| emitter.emit(error!(
                        message("at key {}: {}", key, e),
                        primary(value, "bad identifier"),
                    )))?;
                    Ok((key, sp!(value.span => ident)))
                }).collect_with_recovery()
            })
        };
        macro_rules! pop_ident_map {
            ($name:literal) => { parse_idents($name, pop_map($name)) }
        }

        let out = Eclmap {
            language,
            ins_names: pop_ident_map!("ins_names")?,
            ins_signatures: pop_map("ins_signatures"),
            ins_rets: pop_map("ins_rets"),
            gvar_names: pop_ident_map!("gvar_names")?,
            gvar_types: pop_map("gvar_types"),
            timeline_ins_names: pop_ident_map!("timeline_ins_names")?,
            timeline_ins_signatures: pop_map("timeline_ins_signatures"),
        };
        for (key, _) in maps {
            emitter.emit(warning!(
                message("unrecognized section in eclmap: {:?}", key),
                primary(key, "unrecognized section"),
            )).ignore();
        }

        Ok(out)
    }
}

type GatheredSeqmaps = BTreeMap<Sp<String>, BTreeMap<i32, Sp<String>>>;

fn gather_seqmaps(sections: Vec<SeqmapRawSection<'_>>, emitter: &impl Emitter) -> Result<GatheredSeqmaps, ErrorReported> {
    let mut maps = BTreeMap::new();
    for section in sections {
        let cur_map = maps.entry(section.header.sp_map(ToString::to_string)).or_insert_with(BTreeMap::new);

        for (number, value) in section.lines {
            if let Some(previous) = cur_map.insert(number.value, value.sp_map(ToString::to_string)) {
                // FIXME: Technically we should allow this and keep both as aliases.
                //        https://github.com/ExpHP/truth/issues/49
                return Err(emitter.emit(error!(
                    message("in section '{}': duplicate key error for id {}", section.header, number),
                    primary(value, "redefinition"),
                    secondary(previous, "previous definition"),
                )));
            }
        }
    }
    Ok(maps)
}
