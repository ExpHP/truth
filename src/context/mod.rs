//! Structs that carry important global compiler state.

use std::path::PathBuf;

use crate::ident::GensymContext;
use crate::resolve::Resolutions;
use crate::resolve::rib::Rib;

pub use defs::Defs;
pub mod defs;

pub use consts::Consts;
pub mod consts;

pub use crate::diagnostic::DiagnosticEmitter;

/// Context object for the majority of compilation.
///
/// This is a context object that holds a significant portion of the mutable state that is shared between
/// compiler passes (in particular passes that traverse the AST or that convert between the AST and
/// low-level representations).
///
/// It provides some methods for creating definitions and returning [`DefId`]s.  This is partly historical,
/// but also because it's not clear how to move these methods to [`Defs`] where they might conceptually belong.
///
/// # Limitation of scope
///
//   NOTE: To the future me who is about to delete this section from the documentation:
//         Please consider the examples of concessions in the section.
//         Are you absolutely certain you cannot do something similar to achieve your current goal?
//
/// While there is no doubt a great deal of code which depends on this type (or at least on one or more of
/// its fields), there are a number of phases of compilation that are **forbidden** to depend on this type
/// or any of its fields (except [`DiagnosticEmitter`]), just as a matter of principle.  These are:
///
/// * Parsing of text to AST
/// * Formatting of AST to text
/// * Reading binary files to the low-level representation
/// * Writing the low-level representation to a binary file
///
/// There have been numerous instances of things in the past which appeared that they may require breaking
/// this rule, but it has always been found possible to make concessions in favor of keeping this separation.
/// (e.g. [`llir::RawInstr`] holds an args blob so that reading/writing doesn't require signatures.
/// [`crate::passes::resolve_names::assign_res_ids`] allows the parser to not require `Resolutions`, and
/// [`crate::passes::debug::make_idents_unique`] does the same for the formatter)
#[derive(Debug, Clone)]
pub struct CompilerContext<'ctx> {
    pub diagnostics: DiagnosticEmitter,

    /// Catalogues all loaded mapfiles for generating imports.
    mapfiles: Vec<PathBuf>,
    /// Results of name resolution.  Maps [`ResId`]s to [`DefId`]s.
    pub resolutions: Resolutions,
    /// Stores information about [`DefId`]s.
    pub defs: Defs,
    /// For generating identifiers.
    pub gensym: GensymContext,
    /// Cached const values.
    pub consts: Consts,
    /// The initial set of ribs for name resolution, containing names from mapfiles and meta.
    pub initial_ribs: Vec<Rib>,

    /// Ok so... the field that I initially added that needed a lifetime doesn't need one anymore,
    /// but this is here because we might still eventually need a lifetime for arenas, so I don't
    /// want to have to undo the changes that added 'ctx quite just yet.
    ///
    /// FIXME: remove if still unused after a long time
    _lifetime: std::marker::PhantomData<*mut &'ctx ()>, // invariant
}

impl CompilerContext<'_> {
    fn from_diagnostic_emitter(diagnostics: DiagnosticEmitter) -> Self {
        CompilerContext {
            diagnostics,
            mapfiles: Default::default(),
            resolutions: Default::default(),
            defs: Default::default(),
            gensym: Default::default(),
            consts: Default::default(),
            initial_ribs: Default::default(),
            _lifetime: Default::default(),
        }
    }

    /// Create a [`CompilerContext`] that writes diagnostics to the standard error stream.
    pub fn new_stderr() -> Self { Self::from_diagnostic_emitter(DiagnosticEmitter::new_stderr()) }

    /// Create a [`CompilerContext`] that captures diagnostic output which can be recovered
    /// by calling [`DiagnosticEmitter::get_captured_diagnostics`].
    pub fn new_captured() -> Self { Self::from_diagnostic_emitter(DiagnosticEmitter::new_captured()) }
}
