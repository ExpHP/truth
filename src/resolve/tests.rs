use crate::pos::Files;
use crate::parse::Parse;
use crate::fmt::Format;
use crate::error::CompileError;
use crate::eclmap::Eclmap;
use crate::context::CompilerContext;
use crate::ast;

const ECLMAP: &'static str = r#"!eclmap
!gvar_names
100 ALIAS
101 XLIAS
!gvar_types
100 $
101 %
!ins_names
21 alias
"#;

macro_rules! test {
    (
        // Create a test that uses passes::debug::make_idents_unique to append numerical suffixes to each
        // resolvable identifier, and use `insta` to let a human review this output to verify that the
        // suffixes correctly disambiguate each identifier.
        //
        // (ideally this wouldn't require human intervention, but there's issues with equating either
        //  strings (due to formatting) or AST nodes (due to ResId))
        [snapshot]
        $name:ident = <$ty:ty> $source:literal
    ) => {
        #[test]
        fn $name() { assert_snapshot!(resolve_reformat::<$ty>($source).trim()); }
    };

    (
        // Automated form of the [snapshot] test used when all names in the source code are unique.
        // (i.e. the only suffix that would have been added to each ident is `_0`)
        [all_names_unique]
        $name:ident = <$ty:ty> $source:literal
    ) => {
        #[test]
        fn $name() { check_names_unique::<$ty>($source); }
    };

    (
        // Create a compile-fail snapshot test
        [expect_fail($expected:expr)]
        $name:ident = <$ty:ty> $source:literal) => {
        #[test]
        fn $name() { assert_snapshot!(resolve_expect_err::<$ty>($source, $expected).trim()); }
    };

    (
        // Disable a known broken test.  Wuss.
        [disable]
        $($rest:tt)*
    ) => {};
}

fn resolve<A: ast::Visitable + Parse>(text: &str) -> Result<(A, CompilerContext), (Files, CompileError)> {
    let mut files = Files::new();
    let mut ctx = CompilerContext::new();
    ctx.extend_from_eclmap(None, &Eclmap::parse(ECLMAP).unwrap()).unwrap();

    let mut parsed = files.parse::<A>("<input>", text.as_ref()).unwrap().value;
    crate::passes::resolve_names::assign_res_ids(&mut parsed, &mut ctx).unwrap();
    match crate::passes::resolve_names::run(&parsed, &mut ctx) {
        Ok(()) => Ok((parsed, ctx)),
        Err(e) => Err((files, e)),
    }
}

fn resolve_reformat<A: ast::Visitable + Format + Parse>(text: &str) -> String {
    let (mut parsed, ctx) = resolve::<A>(text).unwrap_or_else(|(files, e)| panic!("{}", e.to_string(&files).unwrap()));

    // add suffixes so we can visualize the effects of name resolution
    crate::passes::debug::make_idents_unique::run(&mut parsed, &ctx.resolutions).unwrap();

    crate::fmt::stringify(&parsed)
}

fn check_names_unique<A: ast::Visitable + Format + Parse>(text: &str) {
    let (mut parsed, ctx) = resolve::<A>(text).unwrap_or_else(|(files, e)| panic!("{}", e.to_string(&files).unwrap()));

    // add suffixes so we can visualize the effects of name resolution
    let count_per_ident = crate::passes::debug::make_idents_unique::run(&mut parsed, &ctx.resolutions).unwrap();

    for (ident, count) in count_per_ident {
        if count != 1 {
            let reformatted = crate::fmt::stringify(&parsed);
            panic!("[all_names_unique] failed on ident '{}'.\nResolved output:\n\n{}", ident, reformatted)
        }
    }
}

fn resolve_expect_err<A: ast::Visitable + Parse>(text: &str, expected: &str) -> String {
    let (files, err) = resolve::<A>(text).err().unwrap();
    let err_msg = err.to_string(&files).unwrap();
    assert!(err_msg.contains(expected), "{}", err_msg);
    err_msg
}

// =========================================================================
// Simple scope tests
//
// These don't reuse any names, so there's only one possibility for resolution.
// The goal is to test whether they can be resolved.

test!(
    [all_names_unique]
    local_basic = <ast::Block> r#"{
    int a = 3;
    int b = a + a;
}"#);

test!(
    [all_names_unique]
    func_basic = <ast::Script> r#"
int foo(int x) {
    return x;
}

script script0 {
    int y = 3;
    foo(y);
}"#);

test!(
    [all_names_unique]
    const_basic = <ast::Script> r#"
const int A = 2;

script script0 {
    int x = A;
}"#);

test!(
    [all_names_unique]
    param_basic = <ast::Script> r#"
void foo(int x) {
    return x;
}"#);

test!(
    [all_names_unique]
    reg_alias_basic = <ast::Block> r#"{
    ins_21(ALIAS, XLIAS);
}"#);

test!(
    [all_names_unique]
    ins_alias_basic = <ast::Block> r#"{
    alias(3, 2.4);
}"#);

test!(
    [all_names_unique]
    func_out_of_order = <ast::Script> r#"
script script0 {
    int x = 3;
    foo(x);
}

int foo(int y) {
    return y;
}"#);

test!(
    [all_names_unique]
    const_out_of_order = <ast::Script> r#"
script script0 {
    int x = A;
}

const int A = 2;
"#);

test!(
    [all_names_unique]
    func_scoped_out_of_order = <ast::Block> r#"{
    if (true) {
        int x = foo();
        if (true) {
            int y = foo();
        }

        int foo() { return 5; }
    }
}"#);

test!(
    [all_names_unique]
    const_scoped_out_of_order = <ast::Block> r#"{
    if (true) {
        int x = A;
        if (true) {
            int y = A;
        }

        const int A = 5;
    }
}"#);

test!(
    [expect_fail("in this scope")]
    local_after_scope_end = <ast::Block> r#"{
    if (true) {
        int a = 4;
        int b = a;  // should be OK
    }
    int c = a;  // should fail at `a`
}"#);

test!(
    [expect_fail("in this scope")]
    const_after_scope_end = <ast::Block> r#"{
    if (true) {
        const int a = 4;
        int b = a;  // should be OK
    }
    int c = a;  // should fail at `a`
}"#);

test!(
    [expect_fail("in this scope")]
    func_after_scope_end = <ast::Block> r#"{
    if (true) {
        int foo() { return 4; }
        int b = foo();  // should be OK
    }
    int c = foo();  // should fail at `foo`
}"#);

test!(
    [expect_fail("in this scope")]
    local_using_itself = <ast::Block> r#"{
    int a = a;  // should fail at second `a`
}"#);

test!(
    [all_names_unique]
    local_using_older_sibling = <ast::Block> r#"{
    // NOTE: We might want to forbid this...
    int x = 2, y = x, z = x + y;
}"#);

test!(
    [all_names_unique]
    scoped_using_each_other = <ast::Block> r#"{
    if (true) {
        const int A = 3;
        int foo() { return bar() + A; }
        int bar() { return foo() + B; }
        const int B = 5;
    }
}"#);

test!(
    // checks mixed accesses (consts using funcs and vice versa) and usage
    // of things from an outer scope
    [all_names_unique]
    scoped_using_each_other_ex = <ast::Block> r#"{
    const int OUTER = 3;
    if (true) {
        const int inner() { return OUTER + INNER; }
        const int INNER = outer() + inner();
    }
    const int outer() { return 3; }
}"#);

test!(
    [expect_fail("in this scope")]
    const_scoped_using_local = <ast::Block> r#"{
    int x = 2;
    const int foo = x;  // should fail at `x`
}"#);

test!(
    [expect_fail("in this scope")]
    func_scoped_using_local = <ast::Block> r#"{
    int x = 2;
    int foo() {
        return x;  // should fail at `x`
    }
}"#);

test!(
    [expect_fail("in this scope")]
    const_scoped_using_outer_local = <ast::Block> r#"{
    int x = 2;
    if (true) {
        const int foo = x;  // should fail at `x`
    }
}"#);

test!(
    [expect_fail("in this scope")]
    func_scoped_using_outer_param = <ast::Block> r#"{
    void foo(int a) {
        int bar() {
            return a;  // should fail at `a`
        }
    }
}"#);

test!(
    // FIXME: Remove the [disable] tag once const params are implemented
    [disable]  // known test failure
    [all_names_unique]
    func_scoped_using_outer_param_const = <ast::Block> r#"{
    void foo(const int a) {
        int bar() {
            return a;  // should miraculously succeed!
        }
    }
}"#);

test!(
    // FIXME: Remove the [disable] tag once default arguments are implemented
    [disable]  // known test failure
    [all_names_unique]
    param_using_outer_const = <ast::Block> r#"{
    int bar(int z = a) {
        return z;
    }
    const int a = 3;
}"#);

test!(
    // FIXME: Remove the [disable] tag once default arguments are implemented
    [disable]  // known test failure
    [expect_fail("in this scope")]
    param_using_outer_local = <ast::Block> r#"{
    int a = 3;
    int bar(int z = a) {  // should fail at `a`
        return z;
    }
}"#);

// =========================================================================
// Shadowing or disjoint tests
//
// Here, names will be used multiple times.  It will be important to check
// the numeric IDs appended to each identifier in the snapshot output to see
// that names are resolved correctly.

test!(
    [snapshot]
    local_shadow = <ast::Block> r#"{
    int a = 3;
    if (true) {
        int a = 4;
        int b = a * a;  // should use inner `a`
    }
    int c = a * a;  // should use outer `a`
}"#);

test!(
    [snapshot]
    const_shadow = <ast::Block> r#"{
    const int a = 3;
    if (true) {
        const int a = 4;  // should be different from outer `a`
        int b = a * a;  // should use inner `a`
    }
    int c = a * a;  // should use outer `a`
}"#);

test!(
    [snapshot]
    const_shadows_outer_local = <ast::Block> r#"{
    int a = 3;
    if (true) {
        int b = a * a;  // should use inner `a`
        const int a = 4;
    }
}"#);

test!(
    [snapshot]
    local_shadows_outer_const = <ast::Block> r#"{
    const int a = 3;    // should be a_0
    if (true) {
        int b = a * a;  // should be a_0 * a_0
        int a = 4;      // should be a_1
        int c = a * a;  // should be a_1 * a_1
    }
}"#);

test!(
    [snapshot]
    local_disjoint = <ast::Block> r#"{
    if (true) {
        int a = 4;
        int b = a * a;  // should use inner `a`
    }
    if (true) {
        int a = 4;  // should be different from other inner `a`
        int b = a * a;  // should use new inner `a`
    }
}"#);

test!(
    [expect_fail("redefinition")]
    local_redefinition = <ast::Block> r#"{
    if (true) {
        int a = 4;
        int b = 3, a = 5;  // should fail on `a`
    }
}"#);

test!(
    [expect_fail("redefinition")]
    func_redefinition = <ast::Script> r#"
int foo(int x) {
    return x;
}

int foo(float y) {
    return y;
}
"#);

test!(
    [expect_fail("redefinition")]
    const_redefinition = <ast::Script> r#"
const int BLUE = 1;
const int RED = 3;
const int BLUE = RED;
"#);

test!(
    [expect_fail("redefinition")]
    func_scoped_redefinition = <ast::Block> r#"{
    if (true) {
        int foo(int x) {
            return x;
        }

        int foo(float y) {
            return y;
        }
    }
}"#);

test!(
    [expect_fail("redefinition")]
    const_scoped_redefinition = <ast::Block> r#"{
    if (true) {
        const int BLUE = 1;
        const int RED = 3;
        const int BLUE = RED;
    }
}"#);

test!(
    [snapshot]
    local_shadows_reg_alias = <ast::Block> r#"{
    ins_21(ALIAS, 3.0);
    if (true) {
        float ALIAS = 4.0;  // should be different `ALIAS`
        float b = ALIAS;
    }
    ins_21(ALIAS, 3.0);  // should be original `ALIAS`
}"#);

test!(
    [snapshot]
    func_scoped_shadows_ins_alias = <ast::Block> r#"{
    alias(4, 3.0);
    if (true) {
        void alias() {}  // should be different `alias`
        alias();
    }
    alias(4, 3.0);  // should be original `alias`
}"#);

test!(
    [all_names_unique]
    const_shadows_reg_alias = <ast::Script> r#"
const int A = ALIAS;
const int ALIAS = 3;
const int B = ALIAS;
"#);

test!(
    [all_names_unique]
    func_shadows_ins_alias = <ast::Script> r#"
void a() { alias(); }
void alias() {}
void b() { alias(); }
"#);

test!(
    [snapshot]
    separate_namespaces = <ast::Script> r#"
    const int a = a();           // should be a_0 then a_1
    const int a() { return a; }  // should be a_1 then a_0
"#);

// =========================================================================

#[should_panic(expected = "resolved multiple times")]
#[test]
fn panics_on_cloned_res() {
    let mut files = Files::new();
    let mut ctx = CompilerContext::new();
    ctx.extend_from_eclmap(None, &Eclmap::parse(ECLMAP).unwrap()).unwrap();

    let mut def = files.parse::<ast::Stmt>("<input>", b"  int x = 2;  ").unwrap();
    let mut cloned = files.parse::<ast::Stmt>("<input>", b"  x = 3;  ").unwrap();
    crate::passes::resolve_names::assign_res_ids(&mut def, &mut ctx).unwrap();
    crate::passes::resolve_names::assign_res_ids(&mut cloned, &mut ctx).unwrap();

    let block = ast::Block(vec![def, cloned.clone(), cloned]);

    crate::passes::resolve_names::run(&block, &mut ctx).unwrap();
}
