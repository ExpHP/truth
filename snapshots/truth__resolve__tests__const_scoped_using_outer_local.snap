---
source: src/resolve/tests.rs
expression: "resolve_expect_err::<ast::Block>(r#\"{\n    int x = 2;\n    if (true) {\n        const int foo = x;  // should fail at `x`\n    }\n}\"#,\n                                 \"nested const\").trim()"
---
error: cannot use local from outside const
  ┌─ <input>:4:25
  │
2 │     int x = 2;
  │         - defined here
3 │     if (true) {
4 │         const int foo = x;  // should fail at `x`
  │                         ^ used in a nested const
